src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -396,6 +397,50 @@ private void VerifyIsAlive()
                 throw new InvalidOperationException(SR.AssemblyLoadContext_Verify_NotUnloading);
             }
         }
+
+        private static readonly AsyncLocal<AssemblyLoadContext> _asyncLocalCurrentContextualReflectionContext = new AsyncLocal<AssemblyLoadContext>();
+
+        public static AssemblyLoadContext CurrentContextualReflectionContext
+        {
+            get { return _asyncLocalCurrentContextualReflectionContext.Value; }
+        }
+
+        public ContextualReflectionScope EnterContextualReflection()
+        {
+            return new ContextualReflectionScope(this);
+        }
+
+        static public ContextualReflectionScope EnterContextualReflection(Assembly activating)
+        {
+            return activating != null ?
+                GetLoadContext(activating).EnterContextualReflection() :
+                new ContextualReflectionScope(null);
+        }
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public struct ContextualReflectionScope : IDisposable
+        {
+            private bool disposed;
+            private AssemblyLoadContext predecessor;
+
+            public ContextualReflectionScope(AssemblyLoadContext activating)
+            {
+                predecessor = AssemblyLoadContext.CurrentContextualReflectionContext;
+                disposed = false;
+                AssemblyLoadContext._asyncLocalCurrentContextualReflectionContext.Value = activating;
+            }
+
+            public void Dispose()
+            {
+                if(!this.disposed)
+                {
+                    AssemblyLoadContext._asyncLocalCurrentContextualReflectionContext.Value = predecessor;
```

Comment:
Do we want to check that the context is what we expect it to be and throw InvalidOperationException otherwise?

For example, the current implementation lets me to `default(ContextualReflectionScope).Dispose()` and it will just clear whatever is there.

Reply:
> the current implementation lets me to `default(ContextualReflectionScope).Dispose()`

I'm not sure how that could be done. I was assuming if the constructors were not in the facade, it could not be constructed. I suppose if I invert the sense of `_disposed` to `_initialized` a default constructor would not be able to just clear...

Reply:
One can always create uninitialized instance of a struct. It cannot be prevented by making the constructor private or not present.

Reply:
I added the throw.  I also added test case playing with bad usage.